[Frontend] Add observable validation

lillian542 opened the pull request:
**Context:**
We don't have validation that observables are supported yet

**Description of the Change:**
Add a transform that checks if the observables on a tape are supported, and raises an error otherwise. Add it to the overall device TransformProgram.

Two other small things added in this PR:
- a more complete docstring for the preprocess function
- switched the order of operator decomposition vs measurement transforms, because measurement transforms may modify the operators (this should not be true the other way around)

**Benefits:**
We raise a clear error early on if unsupported observables are included in the tape.
By reordering the transforms, we don't add any operators to the tape operations after the `decompose` function is called.

lillian542 commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -911,7 +911,7 @@ def is_leaf(obj):
                 return_values, is_leaf=is_leaf
             )
-            
+
```
```suggestion
            
```

lillian542 commented on frontend/catalyst/jax_tracer.py:
```diff
@@ -109,7 +109,7 @@ def mark_gradient_tracing(method: str):
 
 def _make_execution_config(qnode):
-    """Updates the execution_config object with information about execution. This is 
+    """Updates the execution_config object with information about execution. This is
```
```suggestion
    """Updates the execution_config object with information about execution. This is 
```

lillian542 commented on frontend/test/pytest/test_qnode.py:
```diff
@@ -145,17 +145,19 @@ def grad_circ(x: float):
     assert isinstance(config, qml.devices.ExecutionConfig)
     assert config.gradient_method is None
 
+
```
```suggestion
```

lillian542 commented on frontend/test/pytest/test_qnode.py:
```diff
@@ -145,17 +145,19 @@ def grad_circ(x: float):
 
     assert spy.call_count == 2
-    _, config = spy.spy_return    
+    _, config = spy.spy_return
```
```suggestion
    _, config = spy.spy_return    
```

lillian542 commented on frontend/test/pytest/test_qnode.py:
```diff
@@ -136,7 +136,7 @@ def circ(x):
 
     assert spy.call_count == 1
-    _, config = spy.spy_return    
+    _, config = spy.spy_return
```
```suggestion
    _, config = spy.spy_return    
```

lillian542 commented on frontend/test/pytest/test_qnode.py:
```diff
@@ -123,7 +123,7 @@ def cost_fn(x: float):
 @pytest.mark.parametrize("grad_method", ["adjoint", "parameter-shift"])
 def test_qnode_grad_method_stored_on_execution_config(grad_method, mocker):
-    """Test that the grad_method specified on the qnode is updated on the ExecutionConfig 
+    """Test that the grad_method specified on the qnode is updated on the ExecutionConfig
```
```suggestion
    """Test that the grad_method specified on the qnode is updated on the ExecutionConfig 
```

lillian542 commented on frontend/test/pytest/test_qnode.py:
```diff
@@ -148,6 +148,7 @@ def grad_circ(x: float):
     _, config = spy.spy_return    
     assert config.gradient_method == grad_method
 
+
```
```suggestion
```

lillian542 commented:
[sc-62012]

astralcai commented on frontend/test/pytest/test_verification.py:
```diff
@@ -206,6 +217,96 @@ def f(x: float):
+        with pytest.raises(CompileError, match="RX.*not supported as an observable"):
+            qml.qjit(f)()
+
+    @pytest.mark.parametrize(
```
Maybe add a few cases for nested `Prod` and `SProd`?

lillian542 replied:
added

dime10 commented on frontend/catalyst/programs/verification.py:
```diff
@@ -185,3 +192,57 @@ def _obs_checker(obs):
+            if not qjit_capabilities.native_obs.get(obs.name):
+                return False
+
+            if hasattr(obs, "operands"):
+                return all([_observable_is_supported(o) for o in obs.operands])
+            elif hasattr(obs, "ops"):
+                return all([_observable_is_supported(o) for o in obs.ops])
+            elif hasattr(obs, "base"):
+                return _observable_is_supported(obs.base)
```
Nice, might be useful to add a tiny inline comment which property is from which type of observable

lillian542 replied:
done

dime10 commented on frontend/catalyst/programs/verification.py:
```diff
@@ -185,3 +192,57 @@ def _obs_checker(obs):
+
+
+@transform
+def validate_observables(
```
I wonder if the gradient validation of observables shouldn't be moved here, that way observable validation happens in one place rather than two different places

lillian542 replied:
I would say that the observable validation does happen in one place, and that one place is device `preprocess` - having separate validation transforms here just splits that into smaller, easier to read functions. I think for code clarity (and similarity to the PennyLane convention), a series of transforms added conditionally to the `TransformProgram` is better than one large transform that performs many different checks depending on conditional inputs. 

We've moved away from that convention for operations specifically because with nested tapes there's a performance concern and we want to minimize how many times we iterate through the operations. But since we don't have a similar concern for observables, I'd rather stick to several smaller transforms added in `preprocess`.

dime10 reviewed (changes requested):
Looks great 💯 Just two minor comments

lillian542 commented on frontend/test/pytest/test_verification.py:
```diff
@@ -521,10 +626,15 @@ def cir(x: float):
+            qml.Hamiltonian([2, 3], [qml.X(0), qml.Y(1)]), # hamiltonian
+            qml.ops.Hamiltonian([2, 3], [qml.Y(0), qml.X(1)]), # legacy hamiltonian
+            2 * qml.PauliX(0), # sprod
+            (2 * qml.X(0) @ qml.Y(2)) @ (2*qml.X(3)) @ qml.Y(1), # nested prod+sprod
```
this is the updated behaviour for `validate_observables_adjoint_diff`, previously all operators except single observable and Tensor would have passed the validation (even though we've set `PauliX` to be non-differentiable)

lillian542 commented:
@dime10 @astralcai Since your last review, I realized that the `validate_observables_adjoint_diff` transform that merged in another PR also needs to iterate through the component observables like this (i.e. we don't just want to know if operators of type `Sum` are `differentiable`, but also if all of the Sum's component operands are `differentiable`). I refactored a bit to pull the part that iterates through the components of the observable out, and updated to use it in both transforms. Added a few new test cases for the adjoint diff validation as well. 

So might be worthwhile to take a look at the changes in `verification.py` and `test_verification.py` again.

astralcai commented on frontend/catalyst/programs/verification.py:
```diff
@@ -49,6 +63,35 @@ def _verify_nested(
+
+    if isinstance(obs, (Hamiltonian, CompositeOp, SymbolicOp)):
+
+        _obs_checker(obs)
```
There seems to be some inconsistencies around what `_obs_checker` does. In the doc it is stated that `_obs_checker` should return `True` if the observable is supported, but in reality it does nothing if the observable is supported and raises an error if it isn't, which makes it so that `_verify_observable` would also raise an error instead of returning a `bool`.

lillian542 replied:
Good catch, thanks! I initially had a `stopping_condition` like in PL, but decided to switch to `_obs_checker` to mimic the existing transforms for validation in catalyst instead, and didn't update the documentation to reflect that.

lillian542 replied:
all updated now

lillian542 commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -87,6 +87,7 @@
 
 # pylint: disable=unused-argument,too-many-lines,too-many-statements,too-many-arguments
 
+
```
```suggestion
```

lillian542 commented on frontend/test/pytest/test_verification.py:
```diff
@@ -120,6 +121,16 @@ def null_transform(tape, *args, **kwargs):
+    grad_method = "F"
+
+    def __repr__(self):
+        return f"PauliX2"
```
```suggestion
        return "PauliX2"
```

astralcai reviewed (approved):
LGTM

dime10 commented on frontend/catalyst/programs/verification.py:
```diff
@@ -49,6 +57,39 @@ def _verify_nested(
+    both that the overall observable is supported, and that its component
+    parts are supported."""
+
+    if isinstance(obs, Tensor):
```
Should tensor itself not be checked for support as well?

lillian542 replied:
You would think so, but no. Historically everything supports Tensor, so it's actually not included on hard-coded lists of supported observables, and we just check that its components are supported. 

Try `qml.device("lightning.qubit", wires=3).observables`, for example. It does not include `"Tensor"`, but it does still support `Tensor` 🤷‍♀️ 

Hopefully soon it will be gone and we'll only have `Prod` 🙏

dime10 replied:
Huh weird indeed 😅 I feel like it would still be important to check because ultimately what we want to ensure is that a particular device implemented the `__catalyst__qis__TensorObs` (or rather the `QuantumDevice::TensorObservable` method).

It's true that Lightning also doesn't list it in its toml file, but we might want to remedy this in a followup for consistency 🤔 Should we leave a todo?

lillian542 replied:
I'm hoping `Tensor` will be removed completely in a couple of releases, so not sure how high a priority it is to update this. But no harm including it as a task if there's time! And maybe `Tensor` will stick around longer than I'd like 😅

dime10 replied:
```suggestion
    # TODO: Check whether Tensor is supported by the device.
    if isinstance(obs, Tensor):
```

dime10 commented on frontend/catalyst/programs/verification.py:
```diff
@@ -218,9 +263,40 @@ def _obs_checker(obs):
+    """
+
+    def _obs_checker(obs):
+        if not qjit_capabilities.native_obs.get(obs.name):
+            raise CompileError(
+                f"{m.obs} is not supported as an observable on the '{name}' device with Catalyst"
+            )
```
#828 Made me realize that we also need to include Catalyst's restrictions in the verifier, just like we do for operations. For instance, Catalyst only supports `{expval, var, probs, counts, sample, state}` and no other measurement processes. There are probably restrictions on observables as well. Additionally, only `expval` and `var` currently support observables, other MPs don't. I guess the 1st and 3rd point should be done in an MP checker, rather than the observable checker, but the 2nd point does seem relevant.

lillian542 replied:
I think that should be a separate PR; this one is already decently long, and validation of measurement processes will need to start with determining requirements - we need to figure out what the other restrictions are 😅 So let's let this one be validating that the observables are all supported but the device, and make all MP-related validation another task.

dime10 replied:
MPs for sure, but thank you for adding the observable validation for the catalyst restrictions!

dime10 reviewed (commented):
Thanks @lillian542, looks good although I realized something that might be fairly important to check.

dime10 commented on frontend/catalyst/device/qjit_device.py:
```diff
@@ -95,12 +96,35 @@
+    "Hadamard",
+    "Hermitian",
+    "Hamiltonian",
+    "Projector",
+    "SparseHamiltonian",
+    "Exp",
+    "LinearCombination",
```
These three are not supported I believe, judging by [this function](https://github.com/PennyLaneAI/catalyst/blob/e9c8eccdcedbeeb9b41acd17ddffd36148f61e5c/frontend/catalyst/jax_tracer.py#L667):
```suggestion
    "Exp",
```

lillian542 replied:
`LinearCombination` is a `Sum` and should be handled by the `Sum` check. I don't see `Exp` though, am I missing something?

dime10 replied:
Oh if it just works as a sum great! You're right, it seems `Exp` needs to be excluded as well.

Having tests for all of these would also tell which are(n't) supported :)
